packages/gatsby/src/commands/develop.ts

```diff
@@ -1,613 +1,200 @@
-          chalk.yellow(`is deprecated. Please use`),
-          chalk.cyan(fixMap[api].newName),
-          chalk.yellow(
-            `instead. For migration instructions, see ${fixMap[api].docsLink}\nCheck the following files:`
-          )
-        )
-        console.log()
-        deprecatedLocations[api].forEach(file => console.log(file))
-        console.log()
+      if (!doesConfigChangeRequireRestart(lastConfig, newConfig)) {
+        lastConfig = newConfig
+        return
       }
-    })
-  }
-
-  // compiler.hooks.invalid.tap(`log compiling`, function(...args) {
-  //   console.log(`set invalid`, args, this)
-  // })
 
-  compiler.hooks.watchRun.tapAsync(`log compiling`, function (_, done) {
-    if (webpackActivity) {
-      webpackActivity.end()
+      lastConfig = newConfig
     }
-    webpackActivity = report.activityTimer(`Re-building development bundle`, {
-      id: `webpack-develop`,
-    })
-    webpackActivity.start()
-
-    done()
-  })
 
-  let isFirstCompile = true
-  // "done" event fires when Webpack has finished recompiling the bundle.
-  // Whether or not you have warnings or errors, you will get this event.
-  compiler.hooks.done.tapAsync(`print gatsby instructions`, function (
-    stats,
-    done
-  ) {
-    // We have switched off the default Webpack output in WebpackDevServer
-    // options so we are going to "massage" the warnings and errors and present
-    // them in a readable focused way.
-    const messages = formatWebpackMessages(stats.toJson({}, true))
-    const urls = prepareUrls(
-      program.ssl ? `https` : `http`,
-      program.host,
-      program.port
+    console.warn(
+      `develop process needs to be restarted to apply the changes to ${file}`
     )
-    const isSuccessful = !messages.errors.length
-
-    if (isSuccessful && isFirstCompile) {
-      printInstructions(
-        program.sitePackageJson.name || `(Unnamed package)`,
-        urls
-      )
-      printDeprecationWarnings()
-      if (program.open) {
-        Promise.resolve(openurl(urls.localUrlForBrowser)).catch(() =>
-          console.log(
-            `${chalk.yellow(
-              `warn`
-            )} Browser not opened because no browser was found`
-          )
-        )
-      }
-    }
-
-    isFirstCompile = false
+    io.emit(`develop:needs-restart`, {
+      dirtyFile: file,
+    })
+  })
 
-    if (webpackActivity) {
-      reportWebpackWarnings(stats)
+  process.on(`beforeExit`, async () => {
+    proxy.server.close()
+    await Promise.all([watcher.close(), unlock()])
+  })
 
-      if (!isSuccessful) {
-        const errors = structureWebpackErrors(
-          `develop`,
-          stats.compilation.errors
-        )
-        webpackActivity.panicOnBuild(errors)
-      }
-      webpackActivity.end()
-      webpackActivity = null
-    }
+  process.on(`SIGINT`, async () => {
+    await script.stop(`SIGINT`)
+    process.exit()
+  })
 
-    done()
+  process.on(`SIGTERM`, async () => {
```

Comment:
We added these ourselves instead of using `signal-exit` (like we do in `gatsby-cli`) because it doesn't support async exit handlers. Node _does_ however support async exit handlers.